allensdk/brain_observatory/behavior/eye_tracking_processing.py

```diff
@@ -0,0 +1,219 @@
+    the ellipse formula.
+
+    Parameters
+    ----------
+    df_row : pd.Series
+        A row from an eye tracking dataframe containing either:
+        "cr_width", "cr_height"
+        or
+        "eye_width", "eye_height"
+
+    Returns
+    -------
+    float
+        The elliptical area of the eye or cr in pixels^2
+    """
+    return np.pi * df_row.iloc[0] * df_row.iloc[1]
+
+
+def determine_likely_blinks(eye_areas: pd.Series,
+                            pupil_areas: pd.Series,
+                            outliers: pd.Series,
+                            dilation_frames: int = 2) -> pd.Series:
+    """Determine eye tracking frames which contain likely blinks or outliers
+
+    Parameters
+    ----------
+    eye_areas : pd.Series
+        A pandas series of eye areas.
+    pupil_areas : pd.Series
+        A pandas series of pupil areas.
+    outliers : pd.Series
+        A pandas series containing bool values of outlier rows.
+    dilation_frames : int, optional
+        Determines the number of additional adjacent frames to mark as
+        'likely_blink', by default 2.
+
+    Returns
+    -------
+    pd.Series
+        A pandas series of bool values that has the same length as the number
+        of eye tracking dataframe rows (frames).
+    """
+    blinks = pd.isnull(eye_areas) | pd.isnull(pupil_areas) | outliers
+    if dilation_frames > 0:
+        likely_blinks = ndimage.binary_dilation(blinks,
+                                                iterations=dilation_frames)
+    else:
+        likely_blinks = blinks
+    return pd.Series(likely_blinks)
+
+
+def process_eye_tracking_data(eye_data: pd.DataFrame,
+                              frame_times: pd.Series,
+                              z_threshold: float = 3.0,
+                              dilation_frames: int = 2) -> pd.DataFrame:
+    """Processes and refines raw eye tracking data by adding additional
+    computed feature columns.
+
+    Parameters
+    ----------
+    eye_data : pd.DataFrame
+        A 'raw' eye tracking dataframe produced by load_eye_tracking_hdf()
+    frame_times : pd.Series
+        A series of frame times acquired from a behavior + ophy session
+        'sync file'.
+    z_threshold : float
+        z-score values higher than the z_threshold will be considered outliers,
+        by default 3.0.
+    dilation_frames : int, optional
+        Determines the number of additional adjacent frames to mark as
+        'likely_blink', by default 2.
+
+    Returns
+    -------
+    pd.DataFrame
+        A refined eye tracking dataframe that contains additional information
+        about frame times, eye areas, pupil areas, and frames with likely
+        blinks/outliers.
+
+    Raises
+    ------
+    RuntimeError
+        If the number of sync file frame times does not match the number of
+        eye tracking frames.
+    """
+
+    if len(frame_times) != len(eye_data.index):
```

Comment:
@njmei there are sessions where apparently spurious  edges in the sync file causes there to be a mismatch between the sync count and the video frame count. It's good that you will raise an error in this case. Colin F (I can't find his github handle!) and I were just discussing this separately and he might have some ideas on dealing with this issue.